src/main/scala/ai/tripl/arc/transform/SimilarityJoinTransform.scala

```diff
@@ -0,0 +1,214 @@
+        Left(err :: Nil)
+    }
+  }
+}
+
+case class SimilarityJoinTransformStage(
+    plugin: SimilarityJoinTransform,
+    name: String, 
+    description: Option[String], 
+    leftView: String, 
+    leftFields: List[String], 
+    rightView: String, 
+    rightFields: List[String],
+    outputView: String,
+    persist: Boolean,
+    shingleLength: Int,
+    numHashTables: Int,
+    threshold: Double,
+    caseSensitive: Boolean,
+    partitionBy: List[String], 
+    numPartitions: Option[Int], 
+    params: Map[String, String]
+  ) extends PipelineStage {
+
+  override def execute()(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
+    SimilarityJoinTransformStage.execute(this)
+  }
+}
+
+object SimilarityJoinTransformStage {
+
+  def execute(stage: SimilarityJoinTransformStage)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
+
+    val uuid = UUID.randomUUID.toString
+
+    // split input string into individual characters
+    val regexTokenizer = { new RegexTokenizer()
+      .setInputCol(uuid)
+      .setPattern("")
+      .setMinTokenLength(1)      
+      .setToLowercase(!stage.caseSensitive)
+    }
+
+    // produce ngrams to group the characters
+    val nGram = { new NGram()
+      .setInputCol(regexTokenizer.getOutputCol)
+      .setN(stage.shingleLength)
+    }
+
+    // convert to vector
+    val countVectorizer = { new CountVectorizer()
+      .setInputCol(nGram.getOutputCol)
+    }
+
+    // build locality-sensitive hashing model
+    val minHashLSH = { new MinHashLSH()
+      .setInputCol(countVectorizer.getOutputCol)
+      .setNumHashTables(stage.numHashTables)
+      .setOutputCol("lsh")
+    }
+
+    val pipeline = new Pipeline().setStages(Array(regexTokenizer, nGram, countVectorizer, minHashLSH))
+
+    val transformedDF = try {
+
+      val leftView = spark.table(stage.leftView)   
+      val rightView = spark.table(stage.rightView)   
+
+      // create a string space concatenated field
+      val inputLeftView = leftView.select(
+        col("*"), trim(concat(stage.leftFields.map{ field => when(col(field).isNotNull, concat(col(field).cast(StringType), lit(" "))).otherwise("") }:_*)).alias(uuid)
+      )
+      val inputRightView = rightView.select(
+        col("*"), trim(concat(stage.rightFields.map{ field => when(col(field).isNotNull, concat(col(field).cast(StringType), lit(" "))).otherwise("") }:_*)).alias(uuid)
+      )      
+
+      val pipelineModel = pipeline.fit(inputLeftView)
+
+      val datasetA = pipelineModel.transform(inputLeftView)
+      val datasetB = pipelineModel.transform(inputRightView)
+
+      val leftOutputColumns = leftView.columns.map{columnName => col(s"datasetA.${columnName}")}
+      val rightOutputColumns = rightView.columns.map{columnName => col(s"datasetB.${columnName}")}
+
+      pipelineModel.stages(3).asInstanceOf[MinHashLSHModel]
```

Comment:
Consider searching for the MinHashLSHModel instead of by index in case the position changes in the future